File: ancap-mobile/modules/expo-acp-core/ios/ExpoAcpCoreModule.swift
import ExpoModulesCore

#if canImport(acp_mobile_ffiFFI)
import acp_mobile_ffiFFI
#endif

private func acpNativeAvailabilityMessage() -> String {
  "ACP native core is not linked on iOS yet. Run ancap-mobile/scripts/build-ios-native.ps1 on macOS to stage the UniFFI bindings and package acp_mobile_ffiFFI.xcframework, then rebuild the Expo dev client. Use Import wallet until native iOS wiring is finished."
}

private func requireAcpNativeLinked() throws {
  #if canImport(acp_mobile_ffiFFI)
  return
  #else
  throw Exceptions.NotImplemented(acpNativeAvailabilityMessage())
  #endif
}

#if canImport(acp_mobile_ffiFFI)
private func mapWalletError(_ error: AcpWalletException) -> Exception {
  switch error {
  case .InvalidAddress:
    return Exception(name: "ERR_INVALID_ADDRESS", description: error.localizedDescription)
  case .InvalidMnemonic:
    return Exception(name: "ERR_INVALID_MNEMONIC", description: error.localizedDescription)
  case .InvalidAmount:
    return Exception(name: "ERR_INVALID_AMOUNT", description: error.localizedDescription)
  case .InsufficientFunds:
    return Exception(name: "ERR_INSUFFICIENT_FUNDS", description: error.localizedDescription)
  case .Rpc(let message):
    return Exception(name: "ERR_RPC", description: message)
  case .Internal(let message):
    return Exception(name: "ERR_INTERNAL", description: message)
  @unknown default:
    return Exception(name: "ERR_INTERNAL", description: error.localizedDescription)
  }
}
#endif

public class ExpoAcpCoreModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoAcpCore")

    AsyncFunction("createWallet") { () -> [String: Any] in
      try requireAcpNativeLinked()
      #if canImport(acp_mobile_ffiFFI)
      do {
        let wallet = try acpCreateWalletFfi()
        return [
          "address": wallet.address,
          "mnemonic": wallet.mnemonic,
          "keystoreJson": wallet.keystoreJson,
        ]
      } catch let error as AcpWalletException {
        throw mapWalletError(error)
      }
      #else
      throw Exceptions.NotImplemented(acpNativeAvailabilityMessage())
      #endif
    }

    Function("validateAddress") { (address: String) -> Bool in
      #if canImport(acp_mobile_ffiFFI)
      return acpValidateAddressFfi(address: address)
      #else
      return false
      #endif
    }

    AsyncFunction("addressFromKeystore") { (keystoreJson: String) -> String in
      try requireAcpNativeLinked()
      #if canImport(acp_mobile_ffiFFI)
      do {
        return try acpAddressFromKeystoreFfi(keystoreJson: keystoreJson)
      } catch let error as AcpWalletException {
        throw mapWalletError(error)
      }
      #else
      throw Exceptions.NotImplemented(acpNativeAvailabilityMessage())
      #endif
    }

    AsyncFunction("estimateFeeDefault") { () -> [String: Any] in
      try requireAcpNativeLinked()
      #if canImport(acp_mobile_ffiFFI)
      do {
        let fee = try acpEstimateFeeDefaultFfi()
        return [
          "feeAcp": fee.feeAcp,
          "feeUnits": NSNumber(value: fee.feeUnits),
        ]
      } catch let error as AcpWalletException {
        throw mapWalletError(error)
      }
      #else
      throw Exceptions.NotImplemented(acpNativeAvailabilityMessage())
      #endif
    }

    AsyncFunction("signTransfer") { (rpcUrl: String, keystoreJson: String, toAddress: String, amountAcp: String, feeAcp: String?) -> [String: Any] in
      try requireAcpNativeLinked()
      #if canImport(acp_mobile_ffiFFI)
      do {
        let signed = try acpSignTransferFfi(
          rpcUrl: rpcUrl,
          keystoreJson: keystoreJson,
          toAddress: toAddress,
          amountAcp: amountAcp,
          feeAcp: feeAcp
        )
        return [
          "rawTx": signed.rawTx,
          "txid": signed.txid,
        ]
      } catch let error as AcpWalletException {
        throw mapWalletError(error)
      }
      #else
      throw Exceptions.NotImplemented(acpNativeAvailabilityMessage())
      #endif
    }
  }
}

File: ancap-mobile/modules/expo-acp-core/android/src/main/java/expo/modules/acpcore/ExpoAcpCoreModule.kt
package expo.modules.acpcore

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import uniffi.acp_mobile_ffi.AcpWalletException
import uniffi.acp_mobile_ffi.acpAddressFromKeystoreFfi
import uniffi.acp_mobile_ffi.acpCreateWalletFfi
import uniffi.acp_mobile_ffi.acpEstimateFeeDefaultFfi
import uniffi.acp_mobile_ffi.acpSignTransferFfi
import uniffi.acp_mobile_ffi.acpValidateAddressFfi

private fun mapWalletError(e: AcpWalletException): CodedException {
  val code = when (e) {
    is AcpWalletException.InvalidAddress -> "ERR_INVALID_ADDRESS"
    is AcpWalletException.InvalidMnemonic -> "ERR_INVALID_MNEMONIC"
    is AcpWalletException.InvalidAmount -> "ERR_INVALID_AMOUNT"
    is AcpWalletException.InsufficientFunds -> "ERR_INSUFFICIENT_FUNDS"
    is AcpWalletException.Rpc -> "ERR_RPC"
    is AcpWalletException.Internal -> "ERR_INTERNAL"
  }
  return CodedException(code, e.message ?: "ACP wallet error", e)
}

class ExpoAcpCoreModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoAcpCore")

    AsyncFunction("createWallet") {
      try {
        val w = acpCreateWalletFfi()
        mapOf(
          "address" to w.`address`,
          "mnemonic" to w.`mnemonic`,
          "keystoreJson" to w.`keystoreJson`,
        )
      } catch (e: AcpWalletException) {
        throw mapWalletError(e)
      }
    }

    Function("validateAddress") { address: String ->
      acpValidateAddressFfi(address)
    }

    AsyncFunction("addressFromKeystore") { keystoreJson: String ->
      try {
        acpAddressFromKeystoreFfi(keystoreJson)
      } catch (e: AcpWalletException) {
        throw mapWalletError(e)
      }
    }

    AsyncFunction("estimateFeeDefault") {
      try {
        val fee = acpEstimateFeeDefaultFfi()
        mapOf(
          "feeAcp" to fee.`feeAcp`,
          "feeUnits" to fee.`feeUnits`.toLong(),
        )
      } catch (e: AcpWalletException) {
        throw mapWalletError(e)
      }
    }

    AsyncFunction("signTransfer") { rpcUrl: String, keystoreJson: String, toAddress: String, amountAcp: String, feeAcp: String? ->
      try {
        val signed = acpSignTransferFfi(rpcUrl, keystoreJson, toAddress, amountAcp, feeAcp)
        mapOf(
          "rawTx" to signed.`rawTx`,
          "txid" to signed.`txid`,
        )
      } catch (e: AcpWalletException) {
        throw mapWalletError(e)
      }
    }
  }
}
